pkg/pub_worker/lib/src/bin/pana_wrapper.dart

```diff
@@ -162,50 +165,103 @@ Future<(SdkConfig, SdkConfig)> _detectSdks(Pubspec pubspec) async {
   );
 
   // Choose stable Dart and Flutter SDKs for analysis
-  var dartSdk = dartSdks.firstWhereOrNull((sdk) => !sdk.version.isPreRelease) ??
-      (dartSdks.isNotEmpty ? dartSdks.first : null);
-  var flutterSdk =
+  final installedDartSdk =
+      dartSdks.firstWhereOrNull((sdk) => !sdk.version.isPreRelease) ??
+          (dartSdks.isNotEmpty ? dartSdks.first : null);
+  final installedFlutterSdk =
       flutterSdks.firstWhereOrNull((sdk) => !sdk.version.isPreRelease) ??
           (flutterSdks.isNotEmpty ? flutterSdks.first : null);
 
   final needsNewer = needsNewerSdk(
-          sdkVersion: dartSdk?.version,
+          sdkVersion: installedDartSdk?.version,
           constraint: pubspec.dartSdkConstraint) ||
       needsNewerSdk(
-          sdkVersion: flutterSdk?.version,
+          sdkVersion: installedFlutterSdk?.version,
           constraint: pubspec.flutterSdkConstraint);
 
+  var dartSdkPath = installedDartSdk?.path;
+  var flutterSdkPath = installedFlutterSdk?.path;
   String configKind = 'stable';
   if (needsNewer) {
     configKind = 'preview';
-    dartSdk =
-        dartSdks.firstWhereOrNull((sdk) => sdk.version.isPreRelease) ?? dartSdk;
-    flutterSdk =
-        flutterSdks.firstWhereOrNull((sdk) => sdk.version.isPreRelease) ??
-            flutterSdk;
+    dartSdkPath = await _previewDartSdk() ??
+        dartSdks.firstWhereOrNull((sdk) => sdk.version.isPreRelease)?.path ??
+        dartSdkPath;
+    flutterSdkPath = await _previewFlutterSdk() ??
+        flutterSdks.firstWhereOrNull((sdk) => sdk.version.isPreRelease)?.path ??
+        flutterSdkPath;
   }
 
   return (
     SdkConfig(
-      rootPath: dartSdk?.path,
+      rootPath: dartSdkPath,
       configHomePath: _configHomePath('dart', configKind),
     ),
     SdkConfig(
-      rootPath: flutterSdk?.path,
+      rootPath: flutterSdkPath,
       configHomePath: _configHomePath('flutter', configKind),
     ),
   );
 }
+
+Future<String?> _previewDartSdk() async {
+  final latestBeta = await fetchLatestDartSdkVersion(channel: 'beta');
+  return await _installSdk(
+    sdkKind: 'dart',
+    configKind: 'preview',
+    sdkPath: '/home/worker/dart/preview',
+    version: latestBeta?.version ?? 'master',
+  );
+}
+
+Future<String?> _previewFlutterSdk() async {
+  final archive = await fetchFlutterArchive();
+  return await _installSdk(
+    sdkKind: 'flutter',
+    configKind: 'preview',
+    sdkPath: '/home/worker/flutter/preview',
+    version: archive.latestBeta?.cleanVersion ?? 'master',
+  );
+}
+
+Future<String?> _installSdk({
+  required String sdkKind,
+  required String configKind,
+  required String sdkPath,
+  required String version,
+}) async {
+  if (!_isInsideDocker) {
+    return null;
+  }
+  if (!await Directory(sdkPath).exists()) {
+    final configHomePath = _configHomePath(sdkKind, configKind);
+    await runConstrained(
+      [
+        'tool/setup-$sdkKind.sh',
+        sdkPath,
```

Comment:
this is probably not reliable.

Like, we should download with retries, probably, and possibly consider checking the `x-goog-hash` header, like crc32.. This could be a follow up, but worth considering.